File: python-legacy/app/api/endpoints.py
"""Typed fetch functions for the Odds API."""

from __future__ import annotations

import asyncio
import logging

from app.api.client import OddsAPIClient
from app.api.models import Event, Score, Sport

log = logging.getLogger(__name__)


async def get_sports(client: OddsAPIClient) -> list[Sport]:
    """Fetch all available sports (free endpoint)."""
    data = await client.get_free("/sports")
    return [Sport(**s) for s in data]


async def get_odds(
    client: OddsAPIClient,
    sport: str,
    *,
    regions: str = "us",
    markets: str = "h2h,spreads,totals",
    odds_format: str = "american",
    bookmakers: list[str] | None = None,
) -> list[Event]:
    """Fetch odds for a sport. Costs credits."""
    params: dict = {
        "regions": regions,
        "markets": markets,
        "oddsFormat": odds_format,
    }
    if bookmakers:
        params["bookmakers"] = ",".join(bookmakers)
    data = await client.get(f"/sports/{sport}/odds", params=params)
    return [Event(**e) for e in data]


async def get_scores(
    client: OddsAPIClient,
    sport: str,
    *,
    days_from: int = 1,
) -> list[Score]:
    """Fetch live & recent scores. Costs 1 credit per request."""
    params = {"daysFrom": days_from}
    data = await client.get(f"/sports/{sport}/scores", params=params)
    return [Score(**s) for s in data]


async def get_events(client: OddsAPIClient, sport: str) -> list[dict]:
    """Fetch events for a sport (free endpoint)."""
    data = await client.get_free(f"/sports/{sport}/events")
    return data


async def get_event_odds(
    client: OddsAPIClient,
    sport: str,
    event_id: str,
    *,
    regions: str = "us",
    markets: str = "player_pass_yds",
    odds_format: str = "american",
    bookmakers: list[str] | None = None,
) -> Event:
    """Fetch odds for a single event. Costs credits."""
    params: dict = {
        "regions": regions,
        "markets": markets,
        "oddsFormat": odds_format,
    }
    if bookmakers:
        params["bookmakers"] = ",".join(bookmakers)
    data = await client.get(
        f"/sports/{sport}/events/{event_id}/odds", params=params
    )
    return Event(**data)


async def get_props_for_events(
    client: OddsAPIClient,
    sport: str,
    event_ids: list[str],
    *,
    regions: str = "us",
    markets: str = "player_points",
    odds_format: str = "american",
    bookmakers: list[str] | None = None,
    max_concurrent: int = 5,
) -> list[Event]:
    """Fetch prop odds for multiple events concurrently.

    Resilient to individual failures — returns only successful results.
    """
    sem = asyncio.Semaphore(max_concurrent)

    async def _fetch_one(eid: str) -> Event | None:
        async with sem:
            try:
                return await get_event_odds(
                    client,
                    sport,
                    eid,
                    regions=regions,
                    markets=markets,
                    odds_format=odds_format,
                    bookmakers=bookmakers,
                )
            except Exception as exc:
                log.warning("Failed to fetch props for event %s: %s", eid, exc)
                return None

    results = await asyncio.gather(*[_fetch_one(eid) for eid in event_ids])
    return [e for e in results if e is not None]

File: src-tauri/src/api/endpoints.rs
use crate::api::client::OddsApiClient;
use crate::errors::AppResult;
use crate::models::{Event, Score, Sport};
use futures::stream::{self, StreamExt};

#[derive(Debug, Clone)]
pub struct OddsQuery<'a> {
    pub regions: &'a str,
    pub markets: &'a str,
    pub odds_format: &'a str,
    pub bookmakers: Option<&'a [String]>,
}

impl<'a> Default for OddsQuery<'a> {
    fn default() -> Self {
        Self {
            regions: "us",
            markets: "h2h,spreads,totals",
            odds_format: "american",
            bookmakers: None,
        }
    }
}

impl<'a> OddsQuery<'a> {
    fn to_params(&self) -> Vec<(&'static str, String)> {
        let mut v = vec![
            ("regions", self.regions.to_string()),
            ("markets", self.markets.to_string()),
            ("oddsFormat", self.odds_format.to_string()),
        ];
        if let Some(books) = self.bookmakers {
            if !books.is_empty() {
                v.push(("bookmakers", books.join(",")));
            }
        }
        v
    }
}

pub async fn get_sports(client: &OddsApiClient) -> AppResult<Vec<Sport>> {
    client.get("/sports", &[]).await
}

pub async fn get_odds(
    client: &OddsApiClient,
    sport: &str,
    q: OddsQuery<'_>,
) -> AppResult<Vec<Event>> {
    let path = format!("/sports/{sport}/odds");
    let params = q.to_params();
    let params_ref: Vec<(&str, String)> =
        params.iter().map(|(k, v)| (*k, v.clone())).collect();
    client.get(&path, &params_ref).await
}

pub async fn get_scores(
    client: &OddsApiClient,
    sport: &str,
    days_from: u32,
) -> AppResult<Vec<Score>> {
    let path = format!("/sports/{sport}/scores");
    let params = [("daysFrom", days_from.to_string())];
    client.get(&path, &params).await
}

pub async fn get_events(
    client: &OddsApiClient,
    sport: &str,
) -> AppResult<Vec<serde_json::Value>> {
    let path = format!("/sports/{sport}/events");
    client.get(&path, &[]).await
}

pub async fn get_event_odds(
    client: &OddsApiClient,
    sport: &str,
    event_id: &str,
    q: OddsQuery<'_>,
) -> AppResult<Event> {
    let path = format!("/sports/{sport}/events/{event_id}/odds");
    let params = q.to_params();
    let params_ref: Vec<(&str, String)> =
        params.iter().map(|(k, v)| (*k, v.clone())).collect();
    client.get(&path, &params_ref).await
}

pub async fn get_props_for_events(
    client: &OddsApiClient,
    sport: &str,
    event_ids: &[String],
    q: OddsQuery<'_>,
    max_concurrent: usize,
) -> Vec<Event> {
    let max = max_concurrent.max(1);
    let results: Vec<Option<Event>> = stream::iter(event_ids.iter().cloned())
        .map(|eid| {
            let q = q.clone();
            async move {
                match get_event_odds(client, sport, &eid, q).await {
                    Ok(event) => Some(event),
                    Err(err) => {
                        tracing::warn!(event_id = %eid, error = %err, "prop fetch failed");
                        None
                    }
                }
            }
        })
        .buffer_unordered(max)
        .collect()
        .await;

    results.into_iter().flatten().collect()
}
